### ISSUE TITLE
Cannot run application using kafka-streams 2.7.0

### ISSUE THREAD

**Issue by en-vee** [spring-cloud/spring-cloud-stream-binder-kafka#1055]
I have built my Spring cloud stream runnable jar using the following in my build.gradle :
`
implementation 'org.apache.kafka:kafka-streams:2.7.0'
`
My Kafka cluster is running using Apache Kafka Streams 2.7.0.

If I explicitly set my kafka-streams library version to 2.6.0, the application works fine.

However, I get the following error when starting my Kafka streams application :

```
2021-03-26T17:28:46.127 [main] ERROR o.s.boot.SpringApplication - MSG=Application run failed
org.springframework.context.ApplicationContextException: Failed to start bean 'streamsBuilderFactoryManager'; nested exception is java.lang.NoSuchFieldError: TRACE
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:181)
        at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:54)
        at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:356)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:155)
        at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:123)
        at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:940)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:591)
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:767)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:326)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1311)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1300)
        at com.company.kafka.Driver.main(Driver.java:9)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:107)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)
Caused by: java.lang.NoSuchFieldError: TRACE
        at org.apache.kafka.streams.StreamsConfig.<clinit>(StreamsConfig.java:760)
        at org.apache.kafka.streams.KafkaStreams.<init>(KafkaStreams.java:581)
        at org.springframework.kafka.config.StreamsBuilderFactoryBean.start(StreamsBuilderFactoryBean.java:316)
        at org.springframework.cloud.stream.binder.kafka.streams.StreamsBuilderFactoryManager.start(StreamsBuilderFactoryManager.java:81)
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:178)
        ... 23 common frames omitted
```

**Comment by garyrussell** [comment]
`RecordingLevel` (where the TRACE field is defined) is in `kafka-clents`; TRACE was added in 2.7.0.

The `kafka-clients` and `kafka-streams` versions must match.

**Comment by garyrussell** [comment]
If you are using the embedded kafka server (provided by spring-kafka-test); there are a bunch of other jars you need to override too: 

https://docs.spring.io/spring-kafka/docs/current/reference/html/#update-deps

**Comment by sobychacko** [comment]
@en-vee Please share your build configuration. As Gary said, there is some version mismatch going on.

**Comment by en-vee** [comment]
@sobychacko  as requested: (note : I am forcing the kafka-streams library to be 2.7.0, hoping it will automatically resolve all other Kafka dependencies to the matching versions)

build.gradle
----------------
`
plugins {
	id 'org.springframework.boot' version '2.4.2'
	id 'io.spring.dependency-management' version '1.0.11.RELEASE'
	id 'java'
	id "nebula.ospackage" version "8.5.6"
}


group = 'com.company'
ext {
	rpmVersion='0.0.1'
}

sourceCompatibility = '1.11'

repositories {
	mavenCentral()
	maven { url 'https://repo.spring.io/milestone' }
}

ext {
	set('springCloudVersion', "2020.0.0")
}

dependencies {
	implementation project(':common')
	implementation 'org.springframework.boot:spring-boot-starter-web'
	implementation 'org.springframework.boot:spring-boot-starter-actuator'
	implementation 'org.apache.kafka:kafka-streams:2.7.0'
	implementation 'org.springframework.cloud:spring-cloud-stream'
	implementation 'org.springframework.cloud:spring-cloud-stream-binder-kafka'
	implementation 'org.springframework.cloud:spring-cloud-stream-binder-kafka-streams'
	implementation 'org.springframework.kafka:spring-kafka'
	testImplementation 'org.springframework.boot:spring-boot-starter-test'
	testImplementation 'org.springframework.kafka:spring-kafka-test'
}

dependencyManagement {
	imports {
		mavenBom "org.springframework.cloud:spring-cloud-dependencies:${springCloudVersion}"
	}
}
`

**Comment by garyrussell** [comment]
Please learn to use GitHub markdown so we don't have to keep editing your posts.

As I said above, you need to use the correct `kafka-clients` version to match `kafka-stream`.

**Comment by sobychacko** [comment]
@en-vee Yep, you are overriding your kafka-streams library to `2.7.0` while using cloud `2020.0.0` which brings the 2.6.x versions of the kafka client libraries. As Gary said, you need to certainly match them.

**Comment by en-vee** [comment]
Thanks @sobychacko and @garyrussell 
That got the application up and also processing messages.
About @sobychacko 's comments about cloud version 2020.0.0.0, how do I ensure I get the latest Kafka libraries without explicitly setting versions to the latest ones ? Do I need to change the cloud version to do that ?

**Comment by garyrussell** [comment]
You cannot; Boot dependency management prescribes the versions of the kafka libraries.

The next Boot version (2.5) will pull in the 2.7.x libraries by default. With Boot 2.4.x (by the way, 2.4.4 is now available), you will get the 2.6.x libraries.

Hence the appendix in the spring-kafka docs that I pointed you to above https://docs.spring.io/spring-kafka/docs/current/reference/html/#update-deps

**Comment by garyrussell** [comment]
Also, the latest cloud version is 2020.0.2.
